### PR TITLE
Make annotations ``SetOnce`` or ``optional`` where feasible

### PR DESCRIPTION
**Pull request by Marenz**
prequesit for #9364

### CODE REVIEW

File: libsolutil/SetOnce.h [hunk 1]
```diff
@@ -79,6 +79,8 @@ class SetOnce
 	/// @throws BadSetOnceAccess when the stored value has not yet been set
 	T const* operator->() const { return std::addressof(**this); }
 
+	/// @return true if a value was assigned
+	bool set() const { return m_value.has_value(); }
```
Comment (cameel) [inline libsolutil/SetOnce.h:83]: I think `hasValue()` or `isSet()` would be a less ambiguous name.

```suggestion
	bool hasValue() const { return m_value.has_value(); }
```

And by the way, maybe adding a `value()` function would be a good idea too?

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -351,6 +351,9 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 	if (_function.overrides() && _function.isFree())
 		m_errorReporter.syntaxError(1750_error, _function.location(), "Free functions cannot override.");
 
+	if (!_function.modifiers().empty() && _function.isFree())
+		m_errorReporter.syntaxError(6261_error, _function.location(), "Free functions cannot have modifiers.");
+
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: This change looks like it doesn't belong here. Is this something that should be in #9364?

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 3]
```diff
@@ -270,18 +283,23 @@ bool ASTJsonConverter::visit(ImportDirective const& _node)
...
 		make_pair("contractDependencies", getContainerIds(_node.annotation().contractDependencies, true)),
 		make_pair("nodes", toJson(_node.subNodes())),
 		make_pair("scope", idOrNull(_node.scope()))
-	});
+	};
+
+	if (_node.annotation().unimplementedDeclarations.has_value())
+		attributes.emplace_back("fullyImplemented", _node.annotation().unimplementedDeclarations->empty());
+	if (!_node.annotation().linearizedBaseContracts.empty())
+		attributes.emplace_back("linearizedBaseContracts", getContainerIds(_node.annotation().linearizedBaseContracts));
```
Comment (cameel) [inline libsolidity/ast/ASTJsonConverter.cpp:318]: What do you think about adding a helper like `makePairIfHasValue()` that would accept `SetOnce<T>` or `optional<T>` and create a pair with `null` if there's no value? `nulls` are supposed to be automatically removed now so that would let us keep the more readable compact code structure here.

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -2746,6 +2770,9 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 			annotation.isPure = true;
 	}
 
+	if (!annotation.isPure.set())
+		annotation.isPure = false;
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp:2781]:
```suggestion
	if (!annotation.isPure.set())
		annotation.isPure = false;
	else
		solAssert(!annotation.isPure, "");
```

Actually, maybe it would be worthwhile to have a helper on `SetOnce` that does this (i.e. either sets a value or verifies that your new value matches)?

File: libsolutil/SetOnce.h (hunk 1 above)
Comment (Marenz) [reply libsolutil/SetOnce.h:83]: I used `isSet()` before and then @chriseth wanted it to be `set()` :laughing:

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp]: Indeed, I was just about to remove that

File: libsolidity/ast/ASTJsonConverter.cpp (hunk 3 above)
Comment (Marenz) [reply libsolidity/ast/ASTJsonConverter.cpp:318]: While we have the code to auto-remove nulls, I think it's better if new code doesn't even add them in the first place.
But we could extend your idea and have a `addIfSet(attributes, name, variable)`

File: libsolidity/analysis/TypeChecker.cpp (hunk 4 above)
Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp:2781]: Well, in this case it doesn't need to match, it just can't be left unset.

File: libsolutil/SetOnce.h (hunk 1 above)
Comment (cameel) [reply libsolutil/SetOnce.h:83]: Oh :) Yeah, names can be pretty subjective. I'd personally vote for `hasValue()` but I remember that @chriseth prefers a convention where there's no verb in the accessor so `set()` is probably more consistent with the rest of the code.

File: libsolidity/ast/ASTJsonConverter.cpp (hunk 3 above)
Comment (cameel) [reply libsolidity/ast/ASTJsonConverter.cpp:318]: Right. Actually I prefer not adding nulls too but I suggested it anyway because these hairy `if`s bother me more ;) `addIfSet()` might be a good compromise.

File: libsolidity/analysis/TypeChecker.cpp (hunk 4 above)
Comment (cameel) [reply libsolidity/analysis/TypeChecker.cpp:2781]: Oh, ok then. I did not realize that. To be honest these assignments are pretty hard to review without just reading the whole code and reviewing all possible paths.

File: libsolutil/SetOnce.h (hunk 1 above)
Comment (Marenz) [reply libsolutil/SetOnce.h:83]: or we use the same that `optional` has and use `has_value` that way `SetOnce` and `optional` can be used in a more similar way, avoiding confusion.

Comment (cameel) [reply libsolutil/SetOnce.h:83]: Using a different convention for our stuff has the nice (though unintended) side-effect of making it easy to see what was added by us and what's built in. Not sure if that matters much in this simple case, but if I saw `has_value()`, I'd assume that it's probably `optional`, not `SetOnce`.

File: libsolidity/analysis/TypeChecker.cpp (hunk 4 above)
Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp:2781]: Yeah, that part was a bit messy, indeed.

File: libsolidity/ast/ASTJsonConverter.cpp (hunk 3 above)
Comment (Marenz) [reply libsolidity/ast/ASTJsonConverter.cpp:318]: I replaced it now

File: libsolutil/SetOnce.h (hunk 1 above)
Comment (chriseth) [reply libsolutil/SetOnce.h:83]: It's mostly about `is` being redundant in functions returning bool.

File: libsolidity/ast/ASTAnnotations.h [hunk 5]
```diff
@@ -149,7 +150,7 @@ struct StructDeclarationAnnotation: TypeDeclarationAnnotation
 struct ContractDefinitionAnnotation: TypeDeclarationAnnotation, StructurallyDocumentedAnnotation
 {
 	/// List of functions and modifiers without a body. Can also contain functions from base classes.
-	std::vector<Declaration const*> unimplementedDeclarations;
+	std::optional<std::vector<Declaration const*>> unimplementedDeclarations;
```
Comment (chriseth) [inline libsolidity/ast/ASTAnnotations.h:153]: Why optional and not setonce?

File: libsolidity/ast/ASTAnnotations.h [hunk 6]
```diff
@@ -245,14 +246,14 @@ struct ExpressionAnnotation: ASTAnnotation
 	/// Whether the expression is a constant variable
 	bool isConstant = false;
```
Comment (chriseth) [inline libsolidity/ast/ASTAnnotations.h]: Why not this one?

File: libsolidity/ast/ASTAnnotations.h (hunk 5 above)
Comment (Marenz) [reply libsolidity/ast/ASTAnnotations.h:153]: Because it's filled up in 2 different functions.. though if I can actually initialize it once and add content then `SetOnce` would work

File: libsolidity/ast/ASTAnnotations.h (hunk 6 above)
Comment (Marenz) [reply libsolidity/ast/ASTAnnotations.h]: If I recall correctly, it was quite cumbersome setting this in all the visitors correctly. I can have another look though.

Comment (chriseth) [reply libsolidity/ast/ASTAnnotations.h]: Ah right, I remember. It's fine!

File: libsolidity/analysis/ContractLevelChecker.cpp [hunk 7]
```diff
@@ -51,6 +51,8 @@ bool hasEqualNameAndParameters(T const& _a, B const& _b)
 
 bool ContractLevelChecker::check(ContractDefinition const& _contract)
 {
+	_contract.annotation().unimplementedDeclarations = std::vector<Declaration const*>();
```
Comment (chriseth) [inline libsolidity/analysis/ContractLevelChecker.cpp:54]: Would it be possible to change the checking functions to return the unimplemented declarations instead?

Comment (chriseth) [reply libsolidity/analysis/ContractLevelChecker.cpp:54]: Ah ok, maybe just leave it...

File: libsolidity/analysis/TypeChecker.cpp [hunk 8]
```diff
@@ -2325,6 +2335,8 @@ bool TypeChecker::visit(FunctionCallOptions const& _functionCallOptions)
 
 	_functionCallOptions.expression().accept(*this);
 
+	_functionCallOptions.annotation().isPure = *_functionCallOptions.expression().annotation().isPure;
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Was it like that before? What about `f{value: functionWithSideEffects()}()`?

File: libsolidity/analysis/TypeChecker.cpp [hunk 9]
```diff
@@ -2673,11 +2686,16 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 				functionType &&
 				functionType->kind() == FunctionType::Kind::Declaration
 			)
-				annotation.isPure = _memberAccess.expression().annotation().isPure;
+				annotation.isPure = *_memberAccess.expression().annotation().isPure;
 		}
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:2701]: Should this have
```
else annotation.isLValue = false;
```
?

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:2701]: Ah, it's done at the end, right?

File: test/libsolidity/ASTJSON/assembly/nested_functions.json [hunk 10]
```diff
@@ -16,7 +16,6 @@
       "baseContracts": [],
       "contractDependencies": [],
       "contractKind": "contract",
-      "fullyImplemented": true,
```
Comment (chriseth) [inline test/libsolidity/ASTJSON/assembly/nested_functions.json:19]: Why are these removed?

Comment (Marenz) [reply test/libsolidity/ASTJSON/assembly/nested_functions.json:19]: because nested_functions.sol doesn't successfully compile
it does however successfully output an AST tree
but due to the error the analysis is interrupted

File: libsolidity/analysis/TypeChecker.cpp (hunk 8 above)
Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp]: Hmm yeah fair point, I don't know why I didn't just set it to `false`

**Comment by Marenz** [comment]
Pushed changes:
* `isConstant` is now also optional
* removed the functioncalloptions isPure setting which modified logic without intending so

File: libsolidity/ast/ASTAnnotations.h (hunk 5 above)
Comment (Marenz) [reply libsolidity/ast/ASTAnnotations.h:153]: but `SetOnce` only allows `const` access so it has to be `optional`

**Review by chriseth — APPROVED** [review]
